materialize-bigquery/client.go

```diff
@@ -38,6 +38,69 @@ func newClient(ctx context.Context, ep *sql.Endpoint) (sql.Client, error) {
 }
 
 func (c *client) InfoSchema(ctx context.Context, resourcePaths [][]string) (*boilerplate.InfoSchema, error) {
+	// First check if there are any interrupted column migrations which must be resumed, before we
```

Comment:
I don't like having this resumption logic in `InfoSchema`, which is probably not too terribly surprising. I'm also pretty uncomfortable with relying on specific error messages from this systems so much in general...they absolutely do change (I believe some of our prereq checks aren't working super well because of that actually), and there's a potential for data loss / things to get pretty messed up here.

Also I'd prefer for us to not have to create additional metadata tables. Thinking about all the non-transactional checks for existing of / creation of the migration table & updates to it in the context of racing / zombie invocations of the connector is kind of difficult for me to think through totally.

I think the majority of the heavy lifting of actually shuffling columns and data around is done here, but I'm going to suggest a somewhat alternate strategy for resumption:

The idea is to use the existence of the columns as an indicator that a migration was partially complete. This information is available in the `InfoSchema` already. It would all be implemented in the `materialize-sql` `Applier`, in `UpdateResource` specifically. The steps would be:
- `UpdateResource` recognizes that a column migration is needed, and in the happy path starts doing that: Creates a new column, moves data, renames, etc.
   - It should check for the pre-existence of the `_flowtmp1` column first, and if it's already there it doesn't need to be added again, but the data copying step should alway start from the beginning, casting values into `_flowtmp1` and not assuming a prior temp was successful at doing that
 - `UpdateResource` might get told to create a new column if a prior migration got to the point where it renamed the original column, but failed when trying to rename the "new" column, in `BindingUpdate.NewProjections`. It needs to check and see if the `_flowtmp1` column exists before doing that: If it does, instead of adding a new column it should finish renaming the `_flowtmp1` column. As an extra sanity check here, it can also verify that `_flowtmp2` exists.
- The adding, dropping, and renaming of columns _might_ be able to be done using standard SQL syntax and the dialect's resolved identifiers and DDLs. The specific cast statement is probably something that will need to come from a new method on the `sql.Client`.
- If `UpdateResource` sees that there are `_flowtmp*` columns for existing columns that do not require migration, it should clean them up.

So there are a few failure modes that can happen:
1. Connector crash after creating `_flowtmp1`: `_flowtmp1` is seen to exist on the next `Apply` invocation and the migration starts over
2. Connector crash after copying data to `_flowtmp1` but before renaming the original column: Same resumption as above, with the data copied over again
3. Connector crash after renaming the original column: `UpdateResource` will get a `BindingUpdate.NewProjection` for the original column since it does not exist. It will see that `_flowtmp1` is there (and make sure that `_flowtmp2` is also) and proceed by renaming `_flowtmp1` instead of creating an entirely new column
4. Connector crash after renaming the new column: `UpdateResource` always gets called for every apply for every resource, so it can check that there are no vestigial `_flowtmp*` columns and clean them up if needed.

Possible race conditions:
- Concurrent attempts to create `_flowtmp1`: Only one will succeed, others will crash
- Concurrent attempts to copy data into `_flowtmp1`: This can happen, but shouldn't be an issue
- Concurrent attempts to rename the original column: Only one will succeed, otherwise will crash
- Concurrent attempts to rename `_flowtmp1`: Only one will succeed, others will crash
- Concurrent attempts to drop `_flowtmp*`: Only one will succeed, others will crash

Inevitably there are difficult to reason about cases where a migration is partially completed, and the source field schema changes even further with any strategy. I'm imagining a scenario like this:
1. Column needs migrated from `datetime` to `string`
2. Migration crashes after renaming the original column
3. Before the migration resumes, the source field schema gets changed to something else like `[string, bool]` that would require a `JSON` column
4. The next `Apply` renames the `original_tmp1` column to `original`, and then we'll get an error when the materialization tries to put a `bool` into a `string` column. The `Validate` RPC will have snuck through, since the original column wasn't even there briefly.

Perhaps as future work we could enhance `Validate` to actually detect these partial migrations. It could actually be relatively straightforward for `Validate` to at least see that a partial migration is in progress via the existence of the `_tmp*` columns and return an error to prevent this from happening. I don't know that we _need_ to have that ability right away, but it would be nice if it wasn't too big of an add.

Reply:
> I don't like having this resumption logic in InfoSchema, which is probably not too terribly surprising. I'm also pretty uncomfortable with relying on specific error messages from this systems so much in general

I agree

> The idea is to use the existence of the columns as an indicator that a migration was partially complete.

I like this suggestion as well

However I do not really like the idea of involving this process _after_ Validate, I think it brings in a new scope in: Validate and Apply now need to together know that some of the fields they see are not _real_ fields and are temporary fields which need to be migrated, etc., and you mentioned one such issue that can arise if we let Validate accept this (conflict with another migration that is supposed to happen at the same time).

I would prefer to keep the migration resumption as a step before Validate, so that whatever Validate and Apply see are the clean state of things. Then they can reason about them as they currently do, and as such there will not be a problem if another migration is requested as the connector restarts while a migration resumption is pending.

I don't like the current factoring either, so thinking about it again I think what could work is as follows:

- Introduce a new method on `Client` which is essentially an init step where a client can take care of migrations such as this, let's call it "init" for now.
- This init method is responsible to find the `_flowtmp1` and `_flowtmp2` columns (I imagine it querying columns first to find these columns, and then finding their respective table and schema, basically inverse of what `InfoSchema` does)
- Once these columns are found, migration is carried out as you described
- Init method is done, Validate and Apply continue as before with no modifications necessary

This puts more into init because it has to find the columns as well, but it keeps the dirty work of resuming the migration out of Validate and Apply and I think that makes the whole process easier to reason about, it won't be entangled in Validate and Apply.

Reply:
Given the following two assertions I think that the migration must happen after Validate:

1. A call to the `Validate` RPC should _not_ affect any changes to the destination tables.
2. `Validate` calls need to be as fast as possible, since they are frequently executed in interactive user sessions.

Assuming (1) is true, `Validate` must inevitably have some understanding of the possibility of an in-progress migration. And related to (2), I would expect these migrations to sometimes take quite a while on large tables.

I believe both of the above assertions are true: (1) is an established convention already, and as a user I would not expect that clicking around in the UI or editing a separate task would change the types of my materialized table's columns without an explicit message stating something like that and me clicking `Apply`. (2) is I think pretty self-evident, although I don't know 100% for sure how long the migrations could take.

Broadly it makes sense to me to have the migration happen in `Apply`. The purpose of `Apply` is to converge the table state to be compatible with the applied specification, and the migration is part of that.

One side note: 
> I imagine it querying columns first to find these columns, and then finding their respective table and schema, basically inverse of what InfoSchema does

As a general idea, we can use InfoSchema itself to find out which columns exist. It will contain all the information about the existing columns of tables. I don't think we would need to construct separate queries to find the `_flowtmp*` columns; we could just use InfoSchema directly. For example, by using [HasField](https://github.com/estuary/connectors/blob/5985e037248c8ebceaac660b357e317b50b53723/materialize-boilerplate/info_schema.go#L112) to see if the column exists.

Another side note: Ideally, I think we should return constraint messages for columns that can be migrated stating that they will be migrated, and the user should be in the loop on that if they are making interactive changes.